org.scalaide.worksheet/src/org/scalaide/worksheet/runtime/WorksheetsManager.scala

```diff
@@ -1,35 +1,78 @@
 package org.scalaide.worksheet.runtime
 
-import scala.collection.immutable
+import scala.collection.concurrent
+import scala.ref.WeakReference
 import scala.util.Try
 
+import org.eclipse.core.resources.IResourceChangeEvent
+import org.eclipse.core.resources.IResourceChangeListener
+import org.eclipse.core.resources.ResourcesPlugin
 import org.eclipse.core.runtime.IPath
 import org.scalaide.logging.HasLogger
 import org.scalaide.worksheet.ScriptCompilationUnit
 
 import akka.actor.Actor
 import akka.actor.ActorRef
+import akka.actor.ActorSystem
 import akka.actor.Props
 
 object WorksheetsManager {
   final val ActorName = "worksheet-manager"
   def props(): Props = Props(new WorksheetsManager)
+
+  private object ProjectChangeListener {
+    def apply(system: ActorSystem, worksheetRunners: concurrent.Map[IPath, ActorRef]): IResourceChangeListener =
+      new ProjectChangeListener(WeakReference(system), WeakReference(worksheetRunners))
+
+    private class ProjectChangeListener(system: WeakReference[ActorSystem], worksheetRunners: WeakReference[concurrent.Map[IPath, ActorRef]]) extends IResourceChangeListener {
+      import IResourceChangeEvent._
+      def resourceChanged(event: IResourceChangeEvent): Unit = {
+        val resource = event.getResource()
+        resource.getType match {
```

Comment:
Should this be `event.getType` instead? It seems `PRE_CLOSE` and `PRE_DELETE` match `FOLDER` and `PROJECT`..